2. Extracted SNPs for each  based on increasing positions	awk '$2 == /1/ {print}' sub_maize_genotypes.txt | sort -k3,3n > maize_genotypes_chr1.tx		awk '$2 ~/7/ {print}' sub_teosinte_genotypes.txt | sort -k3,3n > teosinte_genotypes_chr7.txt3. Extracted SNPs for each  based on decreasing positions using a for loop	for i in {1..10}; do awk '$2=='$i'' sub_maize_genotypes.txt | sed 's/?/-/g'| sort -k3,3nr | tee maize_genotypes_chr$i.txt; done    	for i in {1..10}; do awk '$2=='$i'' sub_teosinte_genotypes.txt | sed 's/?/-/g'| sort -k3,3nr | tee teosinte_genotypes_chr$i.txt; 	done4. Extracting SNPs with unknown and multiple  position 	awk '$2 ~ /unknown/ {print}' sub_teosinte_genotypes.txt | sort -k3,3n > unknown_teosinte_genotypes.txt 		awk '$2 ~ /multiple/ {print}' sub_teosinte_genotypes.txt | sort -k3,3n > multiple_teosinte_genotypes.txt		awk '$2 ~ /multiple/ {print}' sub_maize_genotypes.txt | sort -k3,3nr > multiple_maize_genotypes.txt		awk '$2 ~ /unknown/ {print}' sub_maize_genotypes.txt | sort -k3,3nr > unknown_maize_genotypes.txt## my intermediary files are in a sub directory./Processed_files## Individual chromosomes are in respective subdirectories